# А вот и наш первый минипроект!
Прежде чем приступим, немного общей информации о минипроектах. В них вам даются данные и ряд вопросов разной сложности. Работайте с ними как вам удобнее. Один из вариантов – анализировать на JupyterHub, другой – у себя локально в Jupyter. Датасеты доступны по ссылкам на Яндекс Диск и на JupyterHub в папке /shared/homeworks/python_ds_miniprojects.
Датасет для этого минипроекта расположен в папке /shared/homeworks/python_ds_miniprojects/2

В первом мини-проекте вам предстоит проанализировать данные о бронировании отелей. Чуть ниже вы можете ознакомиться со списком вопросов и описанием данных. Для удобства проверки, далее задачи разбиты на стэпы. В некоторых из них также можно встретить небольшие подсказки. Успехов :)

# Задачи
1. Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 
2. Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.
3. Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.
4. На сколько ночей в среднем бронируют отели разных типов?
5. Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?
6. Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов
7. Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?
8. Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?
9. Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

# Описание данных
Имеются следующие переменные:

* Hotel – тип отеля (City Hotel или Resort Hotel)  

* Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным

* Lead time – количество дней, прошедших между датой бронирования и датой прибытия  

* Arrival full date – полная дата прибытия

* Arrival date year – год прибытия  

* Arrival date month – месяц прибытия  

* Arrival date week number – номер недели прибытия

* Arrival date day of month – день прибытия

* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

* Adults – число взрослых

* Children – число детей

* Babies – число младенцев 

* Meal – выбранный тип питания

* Country – страна происхождения клиента

* Reserved room type – тип зарезервированного номера

* Assigned room type – тип полученного номера (может отличаться от забронированного)

* Customer type – тип бронирования

* Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину

* Reservation status date – дата обновления статуса


**1. Импортируйте библиотеку pandas как pd. Прочитайте датасет bookings.csv с разделителем ;. Ссылка на датасет находится в переменной path_to_file.**

**Сохраните прочитанный датасет в bookings.**

**Затем запишите первые 7 строк в переменную bookings_head.**

In [1]:
import pandas as pd
df=pd.read_csv('2_bookings.csv', sep=';')
df.shape

(119390, 21)

**2. Посмотрите на размеры таблицы. Сколько всего в ней столбцов?**

Ответ: 21

**3. К какому типу относится большинство переменных?**

In [3]:
df.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [4]:
df[:7]

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

* float64
* **int64**
* **object**

**4. 💀 Сложное задание! 💀**

**Как вы уже могли заметить, при сборе датасета произошла небольшая ошибка, и названия колонок были записаны в неудобном для дальнейшей работы формате. Исправьте ситуацию, приведя названия столбцов к нижнему регистру и заменив пробелы на знак нижнего подчеркивания.**

Например: Is Canceled -> is_canceled

In [22]:
c1, c2 = [], []
for i in range(df.shape[1]):
    c1.append(df.columns[i])
    c2.append(c1[i].lower().replace(' ','_'))
    df = df.rename(columns = {c1[i]: c2[i]})

In [3]:
df = df.rename(columns=lambda col: col.replace(" ", "_").lower())
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

**5. 💀 Сложное задание! 💀**

**Перейдем к исследованию данных! Пользователи из каких стран совершили наибольшее число успешных бронирований? Бронирование считается успешным, если в дальнейшем не было отменено (переменная is_canceled). В качестве ответа выберите страны, входящие в топ-5.**

In [38]:
df \
    .query('is_canceled==0') \
    .groupby('country') \
    .agg({'is_canceled': 'count'}) \
    .sort_values('is_canceled', ascending=False)[:5]

is_canceled
country             
PRT            21071
GBR             9676
FRA             8481
ESP             6391
DEU             6069

* **ESP**
* PLW
* IRL
* RUS
* **GBR**
* GUY
* MRT
* USA
* **PRT**
* **DEU**
* **FRA**

**6.  На сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel? Запишите полученные значения в пропуски с точностью до 2 знаков после точки.**

In [20]:
dd = df \
    .groupby('hotel') \
    .agg({'stays_total_nights': 'mean'}) \
    .round(2)
dd

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

Ответ: 4.32 в Resort Hotel, 2.98 в City Hotel.

**7. Иногда тип номера, присвоенного клиенту (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.**

**Сколько подобных наблюдений встретилось в датасете?**

*отмена бронирования также считается

In [24]:
dd = df \
    .query('assigned_room_type!=reserved_room_type') \
    .shape
    #.agg({'hotel': 'count'})
#len(dd)
dd[0]

14917

Ответ: 14917

**8. Теперь проанализируйте даты запланированного прибытия (arrival_date_year).**

**На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?**

In [4]:
df \
    .query('arrival_date_year==2016') \
    .groupby('arrival_date_month') \
    .agg({'arrival_date_month': 'count'}) \
    .max()

arrival_date_month    6203
dtype: int64

In [33]:
df.query('arrival_date_year==2017').arrival_date_month.value_counts().idxmax()

'May'

In [5]:
df.groupby('arrival_date_year').arrival_date_month.agg(pd.Series.mode)

arrival_date_year
2015    September
2016      October
2017          May
Name: arrival_date_month, dtype: object

* **В 2016 чаще всего бронировали отель на октябрь**
* В 2016 чаще всего оформляли бронирование на сентябрь
* В 2017 чаще всего бронировали отель на октябрь
* В 2016 чаще всего бронировали отель на май
* **В 2017 чаще всего оформляли бронирование на май**

**9. Сгруппируйте данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017?**

**Hint: для этого сначала примените метод groupby, после чего выберите нужную колонку для дальнейших действий :)**

Например:
```python
df.groupby('column_1')['column_2'].what_to_do()
```

In [102]:
df.query('hotel=="City Hotel"').groupby(['arrival_date_year','arrival_date_month']).is_canceled.sum() \
    .groupby('arrival_date_year') \
    .nlargest(1)
    #.max()    
    #.agg({'arrival_date_month': 'count'})

arrival_date_year  arrival_date_year  arrival_date_month
2015               2015               September             1543
2016               2016               October               1947
2017               2017               May                   2217
Name: is_canceled, dtype: int64

2015 => September

2016 => October

2017 => May

**10. Посмотрите на числовые характеристики трёх колонок: adults, children и babies. Какая из них имеет наибольшее среднее значение?**

In [109]:
df[['adults', 'children', 'babies']].mean().max()
#затем .idxmax

1.8564033838679956

* babies
* **adults**
* children

**11. Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?**

* City hotel – отель находится в городе
* Resort hotel – отель курортный

**В качестве ответа укажите наибольшее среднее total_kids, округлив до 2 знаков после точки.**

In [115]:
df['total_kids'] = df.children+df.babies
df.groupby('hotel').total_kids.mean.round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

Ответ: `0.14`

**12. 💀 Сложное задание! 💀**

**Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитать метрику под названием Churn Rate.**

**Churn rate (отток, коэффициент оттока) – это процент подписчиков (например, на push-уведомления от сайта), которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах.**

**В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов!**

**Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверьте, среди какой группы пользователей показатель оттока выше.** 

**В качестве ответа укажите наибольший % оттока, округленный до 2 знаков после точки (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)**

In [130]:
df['has_kids']=df.total_kids>=1
round(df.groupby('has_kids').is_canceled.sum() / df.groupby('has_kids').is_canceled.count()*100,2)

has_kids
False    37.22
True     34.92
Name: is_canceled, dtype: float64

In [129]:
df['has_kids']=df.total_kids>=1
df[['has_kids', 'is_canceled']].value_counts() / df[['has_kids']].value_counts() * 100

has_kids  is_canceled
False     0              62.778717
          1              37.221283
True      0              65.077154
          1              34.922846
Name: count, dtype: float64

Ответ: `37.22%`